library/src/main/java/com/novoda/downloadmanager/lib/DownloadService.java

```diff
@@ -357,24 +357,34 @@ private boolean updateLocked() {
     }
 
     private List<DownloadBatch> fetchBatches(Collection<DownloadInfo> downloads) {
-        List<DownloadBatch> batches = new ArrayList<>();
         Cursor batchesCursor = resolver.query(Downloads.Impl.BATCH_CONTENT_URI, null, null, null, null);
-        batches.clear();
+        List<DownloadBatch> batches = new ArrayList<>(batchesCursor.getCount());
+        List<Long> forDeletion = new ArrayList<>();
         try {
             int idColumn = batchesCursor.getColumnIndexOrThrow(Downloads.Impl.Batches._ID);
+            int deleteIndex = batchesCursor.getColumnIndex(Downloads.Impl.Batches.COLUMN_DELETED);
```

Comment:
The columns are kind of public already since they're just strings/indexes but there's a queried column validation that happens in the update/insert, it's pretty cray cray